### ISSUE TITLE
Inefficient 2D convolution compared to JAX

### ISSUE THREAD

**Issue by null-pointer-access** [pytorch/pytorch#157334]
### 🐛 Describe the bug

When comparing a 2D depthwise convolution implemented in PyTorch vs. pure JAX/XLA on GPU, I observed that the PyTorch version runs roughly 3× slower and draws substantially more power than the equivalent JAX kernel.

```py
import time
import torch
import jax
import jax.numpy as jnp
import numpy as np

# Pure JAX 2D depthwise convolution
def jax_conv2d(inputs, kernel, bias=None):
    return jax.lax.conv_general_dilated(
        lhs=inputs,                # (batch, H, W, C)
        rhs=kernel,                # (Kh, Kw, 1, C)
        window_strides=(1, 1),
        padding='SAME',
        dimension_numbers=('NHWC','HWIO','NHWC'),
        feature_group_count=inputs.shape[-1]
    )

def main():
    batch_size, dim, H, W = 128, 256, 124, 124
    device = torch.device('cuda')

    # PyTorch setup
    torch_input = torch.randn(batch_size, dim, H, W, device=device)
    dwconv_pt = torch.nn.Conv2d(
        in_channels=dim, out_channels=dim,
        kernel_size=3, stride=1, padding=1,
        groups=dim, bias=True
    ).to(device)

    # JAX setup
    jax.config.update('jax_platform_name', 'gpu')
    np_input = np.random.randn(batch_size, H, W, dim).astype('float32')
    jax_input = jnp.array(np_input)
    # Depthwise: in_per_group=1, out=dim
    kernel = jax.random.normal(
        jax.random.PRNGKey(0), (3, 3, 1, dim)
    ) * 0.1

    # Warm-up
    for _ in range(10):
        dwconv_pt(torch_input)
        jax_conv2d(jax_input, kernel).block_until_ready()

    # Timing
    runs = 50
    torch_time, jax_time = 0.0, 0.0
    for _ in range(runs):
        # PyTorch
        t0 = time.time()
        dwconv_pt(torch_input)
        torch.cuda.synchronize()
        torch_time += time.time() - t0

        # JAX
        t0 = time.time()
        out = jax_conv2d(jax_input, kernel)
        out.block_until_ready()
        jax_time += time.time() - t0

    print(f"PyTorch Conv2d: {torch_time/runs*1000:.1f} ms")
    print(f"JAX Conv2d: {jax_time/runs*1000:.1f} ms")

if __name__ == '__main__':
    main()
```

Output:
```
PyTorch Conv2d: 4.8 ms
JAX Conv2d: 1.7 ms
```

Profiling shows that PyTorch convolution dispatches the conv_depthwise2d_forward_kernel kernel (4853us), whereas JAX/XLA generates the conv2d_c1_k1_nhwc kernel (1352us).

### Versions

```
Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 16.0.6 (++20231112100510+7cbf1a259152-1~exp1~20231112100554.106)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.17 (main, Apr  9 2025, 04:03:39) [Clang 20.1.0 ] (64-bit runtime)
Python platform: Linux-5.14.0-503.23.2.el9_5.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H200
GPU 1: NVIDIA H200
GPU 2: NVIDIA H200
GPU 3: NVIDIA H200
GPU 4: NVIDIA H200
GPU 5: NVIDIA H200
GPU 6: NVIDIA H200
GPU 7: NVIDIA H200

Nvidia driver version: 570.86.15
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               256
On-line CPU(s) list:                  0-255
Vendor ID:                            AuthenticAMD
BIOS Vendor ID:                       Advanced Micro Devices, Inc.
Model name:                           AMD EPYC 9534 64-Core Processor
BIOS Model name:                      AMD EPYC 9534 64-Core Processor                
CPU family:                           25
Model:                                17
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            2
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3718.0659
CPU min MHz:                          1500.0000
BogoMIPS:                             4892.67
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap
Virtualization:                       AMD-V
L1d cache:                            4 MiB (128 instances)
L1i cache:                            4 MiB (128 instances)
L2 cache:                             128 MiB (128 instances)
L3 cache:                             512 MiB (16 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-63,128-191
NUMA node1 CPU(s):                    64-127,192-255
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```

**Comment by albanD** [comment]
What happens if you use channel last format in PyTorch so the memory layout of the inputs is the same?
It is expected for convolutions to be more efficient in this format.

**Comment by Conless** [comment]
> What happens if you use channel last format in PyTorch so the memory layout of the inputs is the same? It is expected for convolutions to be more efficient in this format.

Seems that `torch.nn.Conv2d` does not support channel last format?

https://github.com/pytorch/pytorch/blob/02a9d9095f397f56fe64ef07ced8e7ffed6dcba1/torch/nn/modules/conv.py#L452-L454

**Comment by null-pointer-access** [comment]
> What happens if you use channel last format in PyTorch so the memory layout of the inputs is the same? It is expected for convolutions to be more efficient in this format.

@albanD Seems that `torch.nn.Conv2d` does not support channel last format?

https://github.com/pytorch/pytorch/blob/02a9d9095f397f56fe64ef07ced8e7ffed6dcba1/torch/nn/modules/conv.py#L452-L454

**Comment by albanD** [comment]
The shape doesn't change but the format in memory does. You can do so using `.to(memory_format=torch.channels_last).
In your code sample:
```python
torch_input = torch.randn((batch_size, dim, H, W), device=device).to(memory_format=torch.channels_last)
```

**Comment by Conless** [comment]
> The shape doesn't change but the format in memory does. You can do so using `.to(memory_format=torch.channels_last). In your code sample:
> 
> torch_input = torch.randn((batch_size, dim, H, W), device=device).to(memory_format=torch.channels_last)

@albanD Thanks for the clarification! I'm wondering whether channels_last layout typically offers better throughput than the default (contiguous) layout in most cases? If so, it might be helpful to add some note about this in the `torch.nn.Conv2d` documentation to help developers make more informed decisions!

**Comment by null-pointer-access** [comment]
> The shape doesn't change but the format in memory does. You can do so using `.to(memory_format=torch.channels_last). In your code sample:
> 
> torch_input = torch.randn((batch_size, dim, H, W), device=device).to(memory_format=torch.channels_last)

@albanD Thanks for the clarification! I'm wondering whether channels_last layout typically offers better throughput than the default (contiguous) layout in most cases? If so, it might be helpful to add some note about this in the `torch.nn.Conv2d` documentation to help developers make more informed decisions!

**Comment by albanD** [comment]
My vague recollection is that for some shapes, it does help for the convolution. But other layers, like normalization are slower when using channel last.
So whether you want to use channel last or not depends on your whole network and which layers you're using there, not just the conv2d layer :/
